crates/solver/src/settlement_rater.rs

```diff
@@ -243,19 +246,278 @@ impl SettlementRating for SettlementRater {
-            gas_estimate: simulation.gas_estimate,
-            gas_price: effective_gas_price.clone(),
+            gas_estimate,
+            gas_price: effective_gas_price,
             objective_value,
             score,
             ranking: Default::default(),
         };
         Ok(Rating::Ok(rated_settlement))
     }
 }
+
+/// Contains a subset of the configuration options for the submission of a
+/// settlement, needed for the score calculation.
+pub struct SubmissionConfig {
+    pub strategies: Vec<TransactionStrategyArg>,
+    pub disable_high_risk_public_mempool_transactions: bool,
+}
+
+pub struct ScoreCalculator {
+    score_cap: BigRational,
+    submission_config: SubmissionConfig,
+}
+
+impl ScoreCalculator {
+    pub fn new(
+        score_cap: BigRational,
+        strategies: Vec<TransactionStrategyArg>,
+        disable_high_risk_public_mempool_transactions: bool,
+    ) -> Self {
+        Self {
+            score_cap,
+            submission_config: SubmissionConfig {
+                strategies,
+                disable_high_risk_public_mempool_transactions,
+            },
+        }
+    }
+
+    pub fn cost_fail(&self, gas_cost: &BigRational) -> BigRational {
+        if self
+            .submission_config
+            .strategies
+            .contains(&TransactionStrategyArg::PublicMempool)
+            && !self
+                .submission_config
+                .disable_high_risk_public_mempool_transactions
+        {
+            gas_cost.clone()
+        } else {
+            zero()
+        }
+    }
+
+    pub fn compute_score(
+        &self,
+        objective_value: &BigRational,
+        gas_cost: &BigRational,
+        success_probability: f64,
+    ) -> Result<Score> {
+        ensure!(
+            (0.0..=1.0).contains(&success_probability),
+            "success probability must be between 0 and 1."
+        );
+        let success_probability = BigRational::from_float(success_probability).unwrap();
+        let cost_fail = self.cost_fail(gas_cost);
+        let optimal_score = compute_optimal_score(
+            objective_value.clone(),
+            success_probability,
+            cost_fail,
+            self.score_cap.clone(),
+        )?;
+        let score = big_rational_to_u256(&optimal_score).context("Invalid score.")?;
+        Ok(Score::ProtocolWithSolverRisk(score))
+    }
+}
+
+fn compute_optimal_score(
+    objective: BigRational,
+    probability_success: BigRational,
+    cost_fail: BigRational,
+    score_cap: BigRational,
+) -> Result<BigRational> {
+    tracing::trace!(
+        ?objective,
+        ?probability_success,
+        ?cost_fail,
+        "Computing optimal score"
+    );
+
+    // Computes the amount payed out to solver or penalized from solver given the
+    // second highest score. The optimal bidding is such that in the worst case
+    // (reference score == winning score) the winning solver still breaks even
+    // (profit(winning_score) = 0)
```

Comment:
Let's add that this method is monotonically decreasing (and starts at a positive value - if reference score is 0 you get some non-negative payout) and therefore guaranteed to have exactly one 0